[DYN-1454] Unresolved nodes should attempt to be resolved when a new package is loaded

reddyashish opened the pull request:
### Purpose

This PR is to resolve the dummy nodes once the package is downloaded. 
Task: https://jira.autodesk.com/browse/DYN-1454

Once the package is downloaded, the VM is reset and it reloads all the packages. When this is completed in here: https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCore/Engine/EngineController.cs#L505, we attempt to resolve the dummy node's that are present in the current workspace. The ReloadDummyNodes event is triggered and the model will deserialize each of the dummy nodes. If any of them is resolved by the package that was downloaded, then that dummy node will be replaced in the workspace and all the properties that were present before will be set. If they are not resolved, we just skip them. We disable the executions temporarily when the dummy node is replaced by the new resolved node. Once all the dummy nodes are checked, a final run is triggered here: https://github.com/DynamoDS/Dynamo/blob/master/src/DynamoCore/Graph/Workspaces/HomeWorkspaceModel.cs#L326. 

Once all the dummy nodes are reloaded, a TriggerDependencyRegen is triggered on the Workspace Model from the Dynamo Model. This will regenerate the Dependency table and update the package dependency states on the Workspace Dependency View extension. 

![Reload dummy node new](https://user-images.githubusercontent.com/43763136/65722746-e7236800-e07a-11e9-84a4-ed79e6106a25.gif)


### Declarations

Check these if you believe they are true

- [x] The code base is in a better state after this PR
- [ ] Is documented according to the [standards](https://github.com/DynamoDS/Dynamo/wiki/Coding-Standards)
- [ ] The level of testing this PR includes is appropriate
- [ ] User facing strings, if any, are extracted into `*.resx` files
- [ ] All tests pass using the self-service CI.
- [x] Snapshot of UI changes, if any.
- [ ] Changes to the API follow [Semantic Versioning](https://github.com/DynamoDS/Dynamo/wiki/Dynamo-Versions), and are documented in the [API Changes](https://github.com/DynamoDS/Dynamo/wiki/API-Changes) document.

### Reviewers

@mjkkirschner @aparajit-pratap @QilongTang @scottmitchell

aparajit-pratap commented:
> Once all the dummy nodes are reloaded, a TriggerDependencyRegen is triggered on the Workspace Model from the Dynamo Model. This will regenerate the Dependency table and update the package dependency states on the Workspace Dependency View extension.
Do you need to this? Did that not work previously?

reddyashish commented:
The DependencyRegen was being called as soon as the package is loaded. Then the VM is reset and reloads all the libraries. We have added the Reloading of dummy nodes after the VM is reset so the dependency table was not being updated. So this was needed after we attempt to reload all the dummy nodes.

mjkkirschner commented on src/DynamoCore/Engine/EngineController.cs:
```diff
@@ -38,6 +38,11 @@ public class EngineController : LogSourceBase, IAstNodeContainer, IDisposable
+        /// <summary>
+        /// This event is fired to reload the dummy nodes.
+        /// </summary>
+        public static event Action ReloadDummyNodes;
```
think about this when naming events 

```
Do name events with a verb or a verb phrase.

Do give event names a concept of before and after, using the present and past tense. For example, a close event that is raised before a window is closed would be called Closing and one that is raised after the window is closed would be called Closed.

Do not use Before or After prefixes or suffixes to indicate pre and post events.

Do name event handlers (delegates used as types of events) with the EventHandler suffix.

Do use two parameters named sender and e in event handler signatures.

The sender parameter should be of type Object, and the e parameter should be an instance of or inherit from EventArgs.

Do name event argument classes with the EventArgs suffix.
```

mjkkirschner replied:
why is this static? Is there a good reason?

mjkkirschner replied:
thinking more, I think this should be named to represent the state that is happening, not one single method you are using to the event to call...

so sometihng like `VMLibrariesReset`

reddyashish replied:
Makes sense. Changed the event name accordingly and the event is not static now.

mjkkirschner commented on src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs:
```diff
@@ -188,6 +188,19 @@ internal int CurrentPasteOffset
+        /// <summary>
+        /// This event is to trigger the DependencyRegen on the WorkspaceDependencyview extension.
+        /// </summary>
+        public static event Action TriggerDependencyRegen;
```
I think the event should not be way to signal just that we want to trigger some other method, but more about the state of the workspace - 

like

`DummyNodesResolved`

why is this static- is there a good reason?

reddyashish replied:
Renamed the event name based on the state and made it non-static.

reddyashish replied:
Actually, this has to be static. Because when this event is attached to the current workspace in here (https://github.com/DynamoDS/Dynamo/pull/10011/files#diff-14bfe53b9085e8f1b47772149559d0a7R135), the workspace object is different. So had to keep it static for now.

mjkkirschner replied:
@reddyashish couldn't you just attach this handler when the workspace changes?

mjkkirschner commented on src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs:
```diff
@@ -188,6 +188,19 @@ internal int CurrentPasteOffset
+        /// <summary>
+        /// This method invokes the TriggerDependencyRegen event on the workspace model.
+        /// </summary>
+        public void TriggerDependencyRegenEvent()
```
Usually this would be called - 
OnEventName

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -727,6 +727,13 @@ protected DynamoModel(IStartConfiguration config)
             LibraryServices.MessageLogged += LogMessage;
             LibraryServices.LibraryLoaded += LibraryLoaded;
 
+            EngineController = new EngineController(
```
at first glance this scares me - why do we need to create a new EngineController instance???

reddyashish replied:
Yes, I forgot to remove this, we do not want to create a new EngineController object. I have removed this now.

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1724,81 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
             return true;
         }
 
+        // Reload's the dummy nodes and serializes the new node information once the package resolves it. 
```
this comment is a bit hard to make sense of.

reddyashish replied:
Changed the description.

mjkkirschner commented on src/WorkspaceDependencyViewExtension/WorkspaceDependencyView.xaml.cs:
```diff
@@ -116,6 +116,14 @@ internal void DependencyRegen(WorkspaceModel ws)
+        /// <summary>
+        /// Calls the DependencyRegen function when the event is triggered from the dynamo model.
+        /// </summary>
+        internal void TriggerDepenRegen()
```
whats going on here? Why does this need to live in the view?

reddyashish replied:
This is just a helper method to call the DependencyRegen function.

mjkkirschner commented on test/DynamoCoreTests/PackageDependencyTests.cs:
```diff
@@ -502,5 +503,22 @@ public void PackageDependencyStatechangeTestAfterLoadingPackage()
+            LoadPackage(packageDirectory);
+
+            dummyNodes = CurrentDynamoModel.CurrentWorkspace.Nodes.OfType<DummyNode>();
+            Assert.AreEqual(0, dummyNodes.Count());
```
Can you also assert the value of some node in the graph which depends on the dummy node being resolved?

reddyashish replied:
Yes, asserting the value of the node in the test.

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1724,81 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+            JObject dummyNodeJSON = null;
+            WorkspaceModel currentWorkspace = this.CurrentWorkspace;
+
+            if (currentWorkspace == null || string.IsNullOrEmpty(currentWorkspace.FileName))
```
what happens if this is a customNode workspace? Have you tested that? I think we need a test for that case.

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1724,81 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+
+            String filePath = Path.GetFullPath(currentWorkspace.FileName);
+
+            CustomNodeManager.AddUninitializedCustomNodesInPath(
```
why are you doing this here? What does this have to do with dummyNodes? This will load custom nodes??

reddyashish replied:
Sorry removing this as well. Not needed.

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1724,81 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+                                                               false,
+                                                               CustomNodeManager);
+
+                    if (!(resolvedNode is DummyNode))
```
I think a comment is in order here.

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1724,81 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+                    {
+                        currentWorkspace.SetNodeViewDataOnResolvedNode(dn, resolvedNode);
+
+                        // Disable Run's temporarlity while the dummy node is removed and the resolved node is added. 
```
Disable graph runs temporarily while the dummy node is replaced with the resolved version of that node.

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1724,81 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+                        // Disable Run's temporarlity while the dummy node is removed and the resolved node is added. 
+                        EngineController.DisableRun = true;
+                        currentWorkspace.RemoveAndDisposeNode(dn);
+                        currentWorkspace.AddAndRegisterNode(resolvedNode, false);
```
does this mark the workspace modified (unsaved changes?) should we mark that false before the function returns?

reddyashish replied:
It does mark that the workspace is modified when the node and connectors are added. The * symbol appears on the workspace.

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1724,81 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+                            var startNode = connectorModel.Start.Owner;
+                            var endNode = connectorModel.End.Owner;
+
+                            if (startNode is DummyNode && startNode.Name == resolvedNode.Name)
```
what is this bit about?

reddyashish replied:
When I am adding back the connectors, if the startNode or endNode was a dummy node(and it is being replaced with the resolved node), then the new connector should be added on the resolved node. Since the graph could have many dummy nodes, I need to check the GUID before setting the startNode and endNode for the new connector. I have changed this check the GUID property.

mjkkirschner commented on src/WorkspaceDependencyViewExtension/WorkspaceDependencyView.xaml.cs:
```diff
@@ -124,6 +132,7 @@ public WorkspaceDependencyView(WorkspaceDependencyViewExtension viewExtension,Vi
         {
             InitializeComponent();
             currentWorkspace = p.CurrentWorkspaceModel as WorkspaceModel;
+            WorkspaceModel.TriggerDependencyRegen += TriggerDepenRegen;
```
in general unsubscribe any events that you sub to be safe unless you are sure they will not hold a reference.

reddyashish commented on test/DynamoCoreTests/PackageDependencyTests.cs:
```diff
@@ -502,5 +503,22 @@ public void PackageDependencyStatechangeTestAfterLoadingPackage()
+            Assert.AreEqual(2, dummyNodes.Count());
+
+            // Load the HowickMaker dll and check that the dummy nodes have been resolved.
+            string packageDirectory = Path.Combine(TestDirectory, @"core\packageDependencyTests\HowickMaker");
```
Yes, the dll is added to the test directory and I am loading it from there.

reddyashish commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1724,81 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+                        currentWorkspace.SetNodeViewDataOnResolvedNode(dn, resolvedNode);
+
+                        // Disable Run's temporarlity while the dummy node is removed and the resolved node is added. 
+                        EngineController.DisableRun = true;
```
We do this because for every node and every connector that is added on the workspace, a graph run is triggered and it causes overhead. It doesn't cause any issues but we do not need it. Once all the dummy nodes are resolved, one final run is triggered in the end.

reddyashish commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -727,6 +727,13 @@ protected DynamoModel(IStartConfiguration config)
             LibraryServices.MessageLogged += LogMessage;
             LibraryServices.LibraryLoaded += LibraryLoaded;
 
+            EngineController = new EngineController(
```
Yes, removed this from here.

reddyashish commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1724,81 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+
+                    if (!(resolvedNode is DummyNode))
+                    {
+                        currentWorkspace.SetNodeViewDataOnResolvedNode(dn, resolvedNode);
```
We can do that but we do not want to set these properties if the resolvedNode (that is returned from GetNodeModelForDummyNode function) is also a dummy node. Since we are already checking that in DynamoModel.cs after the nodemodel is returned, it would be one extra step if we want to do it in GetNodeModelForDummyNode().

reddyashish replied:
@aparajit-pratap SetNodeViewDataOnResolvedNode() is now being called from inside GetNodeModelForDummyNode method.

reddyashish commented on src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs:
```diff
@@ -1852,6 +1865,80 @@ protected void Log(string msg, WarningLevel severity)
+        /// <param name="isTestMode"></param>
+        /// <param name="verboseLogging"></param>
+        /// <param name="manager"></param>
+        public static NodeModel GetNodeModelForDummyNode(string json, LibraryServices libraryServices,
```
I have moved GetNodeModelForDummyNode and SetNodeViewDataOnResolvedNode API's to the DummyNode class as it makes it easier to find them in future improvements.

mjkkirschner commented on src/DynamoCore/Engine/EngineController.cs:
```diff
@@ -38,6 +38,11 @@ public class EngineController : LogSourceBase, IAstNodeContainer, IDisposable
         public event AstBuiltEventHandler AstBuilt;
 
+        /// <summary>
+        /// This event is notify that the VMLibraries have been reset.
```
`The event notifies clients that the VMLibraries have been reset.`
@aparajit-pratap is there any other info we want to add to this comment about this state - like - it is now safe to run code?

aparajit-pratap replied:
Yes, we can add that the `VM is now _ready_ to run new code`.

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1719,73 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+
+            // Once all the dummy nodes are reloaded, the DummyNodesReloaded event is invoked and
+            // the Dependency table is regenerated in the WorkspaceDependencyView extension. 
+            currentWorkspace.OnDummyNodesReloaded();
```
@reddyashish I think you should set the hasUnsavedChanges = false here manually and add a test for this - @aparajit-pratap what do you think?

reddyashish replied:
Setting hasUnsavedChanges=false and the workspace changes will be saved automatically as there are no changes to the serialized file. Will add this to the test as well.

aparajit-pratap commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -727,6 +727,8 @@ protected DynamoModel(IStartConfiguration config)
             LibraryServices.MessageLogged += LogMessage;
             LibraryServices.LibraryLoaded += LibraryLoaded;
 
+            EngineController.VMLibrariesReset += ReloadDummyNodes;
```
I think this should be done after the call to `ResetEngineInternal` as a new `EngineController` could be created there, in which case this event will not be subscribed by the new controller.

reddyashish replied:
Moved it.

aparajit-pratap commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1719,73 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+                        currentWorkspace.AddAndRegisterNode(resolvedNode, false);
+
+                        // Adding back the connectors for the resolved node. 
+                        List<ConnectorModel> connectors = dn.AllConnectors.ToList();
```
Can you still access the connectors of the `dn` after disposing it above? How about disposing it after this step to make the sequence of steps more logical?

reddyashish replied:
Yes, I can because because the dummy node is removed from the current workspace but I have the dummy node object before the foreach loop.

aparajit-pratap commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1719,73 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+                return;
+            }
+
+            String filePath = Path.GetFullPath(currentWorkspace.FileName);
```
I don't think `filePath` is used anywhere.

reddyashish replied:
Yes, removing this.

aparajit-pratap commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -1715,6 +1719,73 @@ private void SetPeriodicEvaluation(WorkspaceModel ws)
+
+                if (dummyNodeJSON != null)
+                {
+                    // Deserializing the dummy node and verifying if it is resolved by the downloaded package
```
Agree, it would be worth testing if the dummy nodes are also resolved upon importing a dependency (dll) using `File->Import`, not just package download.

reddyashish replied:
Yes, tested the package import workflow and the dummy nodes are resolved as well upon package import.

mjkkirschner commented on src/DynamoCore/Models/DynamoModel.cs:
```diff
@@ -422,6 +421,7 @@ public IEnumerable<WorkspaceModel> Workspaces
         /// Returns authentication manager object for oxygen authentication.
         /// </summary>
         public AuthenticationManager AuthenticationManager { get; set; }
+        public object DependencyView { get; private set; }
```
@reddyashish what is this for?

reddyashish replied:
This is not needed as well and I should have removed. Thanks for catching it.

reddyashish commented on src/DynamoCore/Graph/Nodes/DummyNode.cs:
```diff
@@ -398,5 +404,76 @@ private XmlElement OriginalXmlNodeContent
+            }
+            else
+            {
+                this.Log(string.Format("This graph has a node with id:{0} and name:{1}, but it could not be resolved",
```
Yes, moved the warning statement to the GetNodeModelForDummyNode(). The resolvedNodeModel cannot be null here so I am just checking for the dummy node

reddyashish commented:
I have a bunch of failures on Self CI but they are passing locally. Looking into it more.

reddyashish commented:
@aparajit-pratap @mjkkirschner The reason for those failures was due to the event in here: https://github.com/DynamoDS/Dynamo/pull/10011/files#diff-14bfe53b9085e8f1b47772149559d0a7R135

I have removed the invocation of DependencyRegen event for now and modified the test to use the dynamo samples and not HowickMaker. Waiting on the self serve results to confirm there are no failures. I will then remove the commented part of the code. Looking into other alternatives in the mean time.

mjkkirschner commented:
@reddyashish this looks like it could have been a memory leak -

reddyashish commented:
Self-serve has passed now. https://master-15.jenkins.autodesk.com/view/DYN/job/DYN-DevCI_Self_Service/424/

Please take a look and this is ready to be merged.

mjkkirschner commented on test/DynamoCoreTests/DummyNodeTests.cs:
```diff
@@ -88,5 +97,29 @@ public void WorkspaceIsReadonly_IfXmlDummyNodePresent()
+            var output = GetPreviewValue("a1aba50a873443f2bfb88480a89e3f36");
+            Assert.IsNull(output);
+
+            // Load the HowickMaker dll and check that the dummy nodes have been resolved.
```
comment is out of date.

reddyashish replied:
Changed it. Thanks Mike.

mjkkirschner commented on src/DynamoCore/Engine/EngineController.cs:
```diff
@@ -38,6 +38,11 @@ public class EngineController : LogSourceBase, IAstNodeContainer, IDisposable
+        /// <summary>
+        /// The event notifies client that the VMLibraries have been reset and the VM is now ready to run the new code. 
+        /// </summary>
+        public static event Action VMLibrariesReset;
```
what do you think about making this internal? Is it useful to anyone but us?

mjkkirschner replied:
do we want to continue supporting it?

reddyashish replied:
I made this internal for now as it is being using only inside the DynamoCore package. Even after we refactor that workflow, we still might need this event.

mjkkirschner commented on src/DynamoCore/Graph/Nodes/DummyNode.cs:
```diff
@@ -398,5 +404,75 @@ private XmlElement OriginalXmlNodeContent
         }
+
+        /// <summary>
+        /// This will return the NodeModel for a dummy node. 
```
this comment is a bit vague - what do you think about something like:
```
Deserializes and returns the nodeModel that is represented by the original content of this DummyNode. If this node cannot be resolved, returns a new DummyNode 
```

mjkkirschner commented on src/DynamoCore/Graph/Nodes/DummyNode.cs:
```diff
@@ -398,5 +404,75 @@ private XmlElement OriginalXmlNodeContent
+                return;
+            }
+
+            resolvedNode.X = dummyNode.X;
```
one thought I have about this is - is there a way to ensure we are transferring all view properties of the dummyNode - so we don't forget to update this if new view properties are added? Something like checking the ExtraNodeViewInfo object for all properties or something? This is not required - but something to think about.

aparajit-pratap commented on test/DynamoCoreTests/DummyNodeTests.cs:
```diff
@@ -24,6 +25,14 @@ protected override void GetLibrariesToPreload(List<string> libraries)
             base.GetLibrariesToPreload(libraries);
         }
 
+        private void LoadPackage(string packageDirectory)
```
I think we can move this function to the base class as it is used in more than on derived class, this one and `PackageDependencyTests`.

reddyashish replied:
Yes, moved it DynamoModelTestBase class.

mjkkirschner commented on src/DynamoCore/Graph/Workspaces/WorkspaceModel.cs:
```diff
@@ -188,6 +188,19 @@ internal int CurrentPasteOffset
         private bool workspaceLoaded;
 
+        /// <summary>
+        /// This event is to notify that the dummy nodes have been resolved. 
```
what about: `This event is raised after the workspace tries to resolve dummyNodes - for example after a new package or library is loaded.`

mjkkirschner commented on test/DynamoCoreTests/DummyNodeTests.cs:
```diff
@@ -88,5 +97,29 @@ public void WorkspaceIsReadonly_IfXmlDummyNodePresent()
         }
+
+        [Test]
+        public void ResolveDummyNodesOnDownloadingPackage()
```
please add a test that does this same workflow *inside* a custom node - IE - the current workspace should be a custom node with dummy nodes in it.

mjkkirschner replied:
Are there more complex tests you can think of that we might be missing? 
* dummy nodes connecting to dummy nodes?
* a test that asserts the view properties are transferred correctly? I don't see that in your test - ie check the name, position etc.

reddyashish replied:
Added the test for dummy nodes inside a custom node workspace

mjkkirschner commented:
@reddyashish - looks pretty solid - a few thoughts
* please fix up the comments a bit to make them more useful to other devs who are unfamiliar with this work
* I think this PR needs more tests.

reddyashish commented:
@mjkkirschner merging this now and will create a followup PR for the view test.

mjkkirschner commented:
@reddyashish go for it.
